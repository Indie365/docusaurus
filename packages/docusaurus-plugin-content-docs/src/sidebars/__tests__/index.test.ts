/**
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

import path from 'path';
import {
  loadSidebars,
  processSidebars,
  DefaultSidebars,
  DisabledSidebars,
} from '../index';
import {
  SidebarItem,
  SidebarItemsGenerator,
  Sidebars,
  UnprocessedSidebars,
  SidebarOptions,
} from '../../types';
import {DefaultSidebarItemsGenerator} from '../generator';

describe('loadSidebars', () => {
  const fixtureDir = path.join(__dirname, '__fixtures__', 'sidebars');
  const options: SidebarOptions = {
    sidebarCollapsed: true,
    sidebarCollapsible: true,
  };
  test('sidebars with known sidebar item type', async () => {
    const sidebarPath = path.join(fixtureDir, 'sidebars.json');
    const result = loadSidebars(sidebarPath, options);
    expect(result).toMatchSnapshot();
  });

  test('sidebars with deep level of category', async () => {
    const sidebarPath = path.join(fixtureDir, 'sidebars-category.js');
    const result = loadSidebars(sidebarPath, options);
    expect(result).toMatchSnapshot();
  });

  test('sidebars shorthand and longform lead to exact same sidebar', async () => {
    const sidebarPath1 = path.join(fixtureDir, 'sidebars-category.js');
    const sidebarPath2 = path.join(
      fixtureDir,
      'sidebars-category-shorthand.js',
    );
    const sidebar1 = loadSidebars(sidebarPath1, options);
    const sidebar2 = loadSidebars(sidebarPath2, options);
    expect(sidebar1).toEqual(sidebar2);
  });

  test('sidebars with category but category.items is not an array', async () => {
    const sidebarPath = path.join(
      fixtureDir,
      'sidebars-category-wrong-items.json',
    );
    expect(() =>
      loadSidebars(sidebarPath, options),
    ).toThrowErrorMatchingInlineSnapshot();
  });

  test('sidebars with category but category label is not a string', async () => {
    const sidebarPath = path.join(
      fixtureDir,
      'sidebars-category-wrong-label.json',
    );
    expect(() =>
      loadSidebars(sidebarPath, options),
    ).toThrowErrorMatchingInlineSnapshot(
      `"Error loading {\\"type\\":\\"category\\",\\"label\\":true,\\"items\\":[\\"doc1\\"]}: \\"label\\" must be a string."`,
    );
  });

  test('sidebars item doc but id is not a string', async () => {
    const sidebarPath = path.join(
      fixtureDir,
      'sidebars-doc-id-not-string.json',
    );
    expect(() =>
      loadSidebars(sidebarPath, options),
    ).toThrowErrorMatchingInlineSnapshot(
      `"Error loading {\\"type\\":\\"doc\\",\\"id\\":[\\"doc1\\"]}: \\"id\\" must be a string."`,
    );
  });

  test('sidebars with first level not a category', async () => {
    const sidebarPath = path.join(
      fixtureDir,
      'sidebars-first-level-not-category.js',
    );
    const result = loadSidebars(sidebarPath, options);
    expect(result).toMatchSnapshot();
  });

  test('sidebars link', async () => {
    const sidebarPath = path.join(fixtureDir, 'sidebars-link.json');
    const result = loadSidebars(sidebarPath, options);
    expect(result).toMatchSnapshot();
  });

  test('sidebars link wrong label', async () => {
    const sidebarPath = path.join(fixtureDir, 'sidebars-link-wrong-label.json');
    expect(() =>
      loadSidebars(sidebarPath, options),
    ).toThrowErrorMatchingInlineSnapshot();
  });

  test('sidebars link wrong href', async () => {
    const sidebarPath = path.join(fixtureDir, 'sidebars-link-wrong-href.json');
    expect(() =>
      loadSidebars(sidebarPath, options),
    ).toThrowErrorMatchingInlineSnapshot();
  });

  test('sidebars with unknown sidebar item type', async () => {
    const sidebarPath = path.join(fixtureDir, 'sidebars-unknown-type.json');
    expect(() =>
      loadSidebars(sidebarPath, options),
    ).toThrowErrorMatchingInlineSnapshot();
  });

  test('sidebars with known sidebar item type but wrong field', async () => {
    const sidebarPath = path.join(fixtureDir, 'sidebars-wrong-field.json');
    expect(() =>
      loadSidebars(sidebarPath, options),
    ).toThrowErrorMatchingInlineSnapshot();
  });

  test('unexisting path', () => {
    expect(loadSidebars('badpath', options)).toEqual(DisabledSidebars);
  });

  test('undefined path', () => {
    expect(loadSidebars(undefined, options)).toEqual(DefaultSidebars);
  });

  test('literal false path', () => {
    expect(loadSidebars(false, options)).toEqual(DisabledSidebars);
  });

  test('sidebars with category.collapsed property', async () => {
    const sidebarPath = path.join(fixtureDir, 'sidebars-collapsed.json');
    const result = loadSidebars(sidebarPath, options);
    expect(result).toMatchSnapshot();
  });

  test('sidebars with category.collapsed property at first level', async () => {
    const sidebarPath = path.join(
      fixtureDir,
      'sidebars-collapsed-first-level.json',
    );
    const result = loadSidebars(sidebarPath, options);
    expect(result).toMatchSnapshot();
  });
});

describe('processSidebars', () => {
  const StaticGeneratedSidebarSlice: SidebarItem[] = [
    {type: 'doc', id: 'doc-generated-id-1'},
    {type: 'doc', id: 'doc-generated-id-2'},
  ];

  const StaticSidebarItemsGenerator: SidebarItemsGenerator = jest.fn(
    async () => {
      return StaticGeneratedSidebarSlice;
    },
  );

  async function testProcessSidebars(unprocessedSidebars: UnprocessedSidebars) {
    return processSidebars({
      sidebarItemsGenerator: StaticSidebarItemsGenerator,
      unprocessedSidebars,
      docs: [],
      // @ts-expect-error: useless for this test
      version: {},
    });
  }

  test('let sidebars without autogenerated items untouched', async () => {
    const unprocessedSidebars: UnprocessedSidebars = {
      someSidebar: [
        {type: 'doc', id: 'doc1'},
        {
          type: 'category',
          collapsed: false,
          collapsible: true,
          items: [{type: 'doc', id: 'doc2'}],
          label: 'Category',
        },
        {type: 'link', href: 'https://facebook.com', label: 'FB'},
      ],
      secondSidebar: [
        {type: 'doc', id: 'doc3'},
        {type: 'link', href: 'https://instagram.com', label: 'IG'},
        {
          type: 'category',
          collapsed: false,
          collapsible: true,
          items: [{type: 'doc', id: 'doc4'}],
          label: 'Category',
        },
      ],
    };

    const processedSidebar = await testProcessSidebars(unprocessedSidebars);
    expect(processedSidebar).toEqual(unprocessedSidebars);
  });

  test('replace autogenerated items by generated sidebars slices', async () => {
    const unprocessedSidebars: UnprocessedSidebars = {
      someSidebar: [
        {type: 'doc', id: 'doc1'},
        {
          type: 'category',
          collapsed: false,
          collapsible: true,
          items: [
            {type: 'doc', id: 'doc2'},
            {type: 'autogenerated', dirName: 'dir1'},
          ],
          label: 'Category',
        },
        {type: 'link', href: 'https://facebook.com', label: 'FB'},
      ],
      secondSidebar: [
        {type: 'doc', id: 'doc3'},
        {type: 'autogenerated', dirName: 'dir2'},
        {type: 'link', href: 'https://instagram.com', label: 'IG'},
        {type: 'autogenerated', dirName: 'dir3'},
        {
          type: 'category',
          collapsed: false,
          collapsible: true,
          items: [{type: 'doc', id: 'doc4'}],
          label: 'Category',
        },
      ],
    };

    const processedSidebar = await testProcessSidebars(unprocessedSidebars);

    expect(StaticSidebarItemsGenerator).toHaveBeenCalledTimes(3);
    expect(StaticSidebarItemsGenerator).toHaveBeenCalledWith({
      defaultSidebarItemsGenerator: DefaultSidebarItemsGenerator,
      item: {type: 'autogenerated', dirName: 'dir1'},
      docs: [],
      version: {},
    });
    expect(StaticSidebarItemsGenerator).toHaveBeenCalledWith({
      defaultSidebarItemsGenerator: DefaultSidebarItemsGenerator,
      item: {type: 'autogenerated', dirName: 'dir2'},
      docs: [],
      version: {},
    });
    expect(StaticSidebarItemsGenerator).toHaveBeenCalledWith({
      defaultSidebarItemsGenerator: DefaultSidebarItemsGenerator,
      item: {type: 'autogenerated', dirName: 'dir3'},
      docs: [],
      version: {},
    });

    expect(processedSidebar).toEqual({
      someSidebar: [
        {type: 'doc', id: 'doc1'},
        {
          type: 'category',
          collapsed: false,
          collapsible: true,
          items: [{type: 'doc', id: 'doc2'}, ...StaticGeneratedSidebarSlice],
          label: 'Category',
        },
        {type: 'link', href: 'https://facebook.com', label: 'FB'},
      ],
      secondSidebar: [
        {type: 'doc', id: 'doc3'},
        ...StaticGeneratedSidebarSlice,
        {type: 'link', href: 'https://instagram.com', label: 'IG'},
        ...StaticGeneratedSidebarSlice,
        {
          type: 'category',
          collapsed: false,
          collapsible: true,
          items: [{type: 'doc', id: 'doc4'}],
          label: 'Category',
        },
      ],
    } as Sidebars);
  });
});
