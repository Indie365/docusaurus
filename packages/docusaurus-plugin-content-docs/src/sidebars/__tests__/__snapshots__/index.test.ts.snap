// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`loadUnprocessedSidebars sidebars link 1`] = `
Object {
  "docs": Array [
    Object {
      "collapsed": true,
      "collapsible": true,
      "items": Array [
        Object {
          "href": "https://github.com",
          "label": "category",
          "type": "link",
        },
      ],
      "label": "Test",
      "type": "category",
    },
  ],
}
`;

exports[`loadUnprocessedSidebars sidebars with category.collapsed property 1`] = `
Object {
  "docs": Array [
    Object {
      "collapsed": true,
      "collapsible": true,
      "items": Array [
        Object {
          "collapsed": false,
          "collapsible": true,
          "items": Array [
            Object {
              "displayThisSidebar": true,
              "id": "doc1",
              "type": "doc",
            },
          ],
          "label": "Introduction",
          "type": "category",
        },
      ],
      "label": "Test",
      "type": "category",
    },
    Object {
      "collapsed": true,
      "collapsible": true,
      "items": Array [
        Object {
          "collapsed": false,
          "collapsible": true,
          "items": Array [
            Object {
              "displayThisSidebar": true,
              "id": "doc2",
              "type": "doc",
            },
          ],
          "label": "Powering MDX",
          "type": "category",
        },
      ],
      "label": "Reference",
      "type": "category",
    },
  ],
}
`;

exports[`loadUnprocessedSidebars sidebars with category.collapsed property at first level 1`] = `
Object {
  "docs": Array [
    Object {
      "collapsed": false,
      "collapsible": true,
      "items": Array [
        Object {
          "displayThisSidebar": true,
          "id": "doc1",
          "type": "doc",
        },
      ],
      "label": "Introduction",
      "type": "category",
    },
    Object {
      "collapsed": false,
      "collapsible": true,
      "items": Array [
        Object {
          "displayThisSidebar": true,
          "id": "doc2",
          "type": "doc",
        },
      ],
      "label": "Powering MDX",
      "type": "category",
    },
  ],
}
`;

exports[`loadUnprocessedSidebars sidebars with deep level of category 1`] = `
Object {
  "docs": Array [
    Object {
      "collapsed": true,
      "collapsible": true,
      "items": Array [
        Object {
          "displayThisSidebar": true,
          "id": "a",
          "type": "doc",
        },
        Object {
          "collapsed": true,
          "collapsible": true,
          "items": Array [
            Object {
              "collapsed": true,
              "collapsible": true,
              "items": Array [
                Object {
                  "displayThisSidebar": true,
                  "id": "c",
                  "type": "doc",
                },
                Object {
                  "collapsed": true,
                  "collapsible": true,
                  "items": Array [
                    Object {
                      "displayThisSidebar": true,
                      "id": "d",
                      "type": "doc",
                    },
                    Object {
                      "collapsed": true,
                      "collapsible": true,
                      "items": Array [
                        Object {
                          "displayThisSidebar": true,
                          "id": "e",
                          "type": "doc",
                        },
                      ],
                      "label": "deeper more more",
                      "type": "category",
                    },
                  ],
                  "label": "level 4",
                  "type": "category",
                },
              ],
              "label": "level 3",
              "type": "category",
            },
            Object {
              "displayThisSidebar": true,
              "id": "f",
              "type": "doc",
            },
          ],
          "label": "level 2",
          "type": "category",
        },
      ],
      "label": "level 1",
      "type": "category",
    },
  ],
}
`;

exports[`loadUnprocessedSidebars sidebars with first level not a category 1`] = `
Object {
  "docs": Array [
    Object {
      "collapsed": true,
      "collapsible": true,
      "items": Array [
        Object {
          "displayThisSidebar": true,
          "id": "greeting",
          "type": "doc",
        },
      ],
      "label": "Getting Started",
      "type": "category",
    },
    Object {
      "displayThisSidebar": true,
      "id": "api",
      "type": "doc",
    },
  ],
}
`;

exports[`loadUnprocessedSidebars sidebars with known sidebar item type 1`] = `
Object {
  "docs": Array [
    Object {
      "collapsed": true,
      "collapsible": true,
      "items": Array [
        Object {
          "displayThisSidebar": true,
          "id": "foo/bar",
          "type": "doc",
        },
        Object {
          "displayThisSidebar": true,
          "id": "foo/baz",
          "type": "doc",
        },
        Object {
          "href": "https://github.com",
          "label": "Github",
          "type": "link",
        },
        Object {
          "displayThisSidebar": false,
          "id": "hello",
          "type": "doc",
        },
      ],
      "label": "Test",
      "type": "category",
    },
    Object {
      "collapsed": true,
      "collapsible": true,
      "items": Array [
        Object {
          "displayThisSidebar": true,
          "id": "hello",
          "type": "doc",
        },
      ],
      "label": "Guides",
      "type": "category",
    },
  ],
}
`;
